x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestFileStructureAction.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.rest;
+
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.xpack.core.ml.action.FileStructureAction;
+import org.elasticsearch.xpack.core.ml.filestructurefinder.FileStructure;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.filestructurefinder.FileStructureFinderManager;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+
+public class RestFileStructureAction extends BaseRestHandler {
+
+    public RestFileStructureAction(Settings settings, RestController controller) {
+        super(settings);
+        controller.registerHandler(RestRequest.Method.POST, MachineLearning.BASE_PATH + "file_structure", this);
+    }
+
+    @Override
+    public String getName() {
+        return "xpack_ml_file_structure_action";
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
+
+        FileStructureAction.Request request = new FileStructureAction.Request();
+        request.setLinesToSample(restRequest.paramAsInt(FileStructureAction.Request.LINES_TO_SAMPLE.getPreferredName(),
+            FileStructureFinderManager.DEFAULT_IDEAL_SAMPLE_LINE_COUNT));
+        if (restRequest.hasContent()) {
+            request.setSample(restRequest.content());
+        } else {
+            throw new ElasticsearchParseException("request body is required");
+        }
+
+        return channel -> client.execute(FileStructureAction.INSTANCE, request, new RestToXContentListener<>(channel));
+    }
+
+    @Override
+    protected Set<String> responseParams() {
+        return Collections.singleton(FileStructure.EXPLAIN);
```

Comment:
This is just for my own edification. 

Does this "magically" take the query parameter the user provides and the passes it to the xcontent builder? I don't see that parameter directly referenced from the incoming request and it has be curious :)

Reply:
All the `RestRequest` parameters are automatically passed to the `toXContent()` method - see 
https://github.com/elastic/elasticsearch/blob/6a3adbd93565299c9e2c09b3781a2d051d22550d/server/src/main/java/org/elasticsearch/rest/action/RestToXContentListener.java#L47

This method just stops the parameter that is _only_ used for X-content formatting and not added to the action request from triggering the validation check for unused parameters.

Reply:
TIL
Thanks for the run down!